File: src/main/cpp/readwrite/pace.cpp
#include "readwrite/pace.hpp"

using namespace ds::graph;

namespace readwrite {
ds::ProblemInstance read_pace(std::istream &is) {
  std::vector<std::pair<Graph::Vertex, std::vector<Graph::Color>>> vertices;
  std::vector<std::pair<Graph::Edge, Graph::Weight>> edges;
  Graph::Vertex s = -1, t = -1;
  std::size_t n = -1, m = -1, c = -1;
  int i = 0, k = -1;

  for (std::string line; std::getline(is, line);) {
    if (line.empty()) continue;

    while (i < static_cast<int>(line.size()) && !std::isprint(line[i])) ++i;
    if (line[i] == 'c') continue;  // ignore comments

    auto tokens = util::split(line);
    if (tokens[0] == "p") {  // p-line
      if (tokens.size() < 5) throw std::invalid_argument(util::format("unexpected p-line: %s", line.c_str()));
      if (tokens[1] != "ip") throw std::invalid_argument(util::format("unexpected p-line: %s", line.c_str()));

      n = std::stoi(tokens[2]);
      m = std::stoi(tokens[3]);
      c = std::stoi(tokens[4]);
      if (tokens.size() >= 6) s = std::stoi(tokens[5]);
      if (tokens.size() >= 7) t = std::stoi(tokens[6]);
      if (tokens.size() >= 8) k = std::stoi(tokens[7]);
    } else if (tokens[0] == "v") {  // v-line
      if (tokens.size() < 1) throw std::invalid_argument("unexpected v-line");

      auto v = std::stoi(tokens[1]);
      std::vector<Graph::Color> colors;
      for (std::size_t i = 2; i < tokens.size(); ++i) colors.push_back(std::stoi(tokens[i]));
      vertices.push_back({v, colors});
    } else if (tokens[0] == "e") {  // e-line
      if (tokens.size() != 4) throw std::invalid_argument("unexpected e-line");

      auto u = std::stoi(tokens[1]);
      auto v = std::stoi(tokens[2]);
      auto w = std::stod(tokens[3]);
      edges.push_back({{u, v}, w});
    }
  }

  ds::graph::Graph G{vertices, edges};
  // verification
  if (n != G.number_of_vertices()) {
    throw std::invalid_argument(util::format("inconsistent n: expect=%d, actual=%lu", n, G.number_of_vertices()));
  }
  if (m != G.number_of_edges()) {
    throw std::invalid_argument(util::format("inconsistent m: expect=%d, actual=%lu", m, G.number_of_edges()));
  }
  if (c != G.number_of_colors()) {
    throw std::invalid_argument(util::format("inconsistent c: expect=%d, actual=%lu", c, G.number_of_colors()));
  }

  ds::ProblemInstance ret = {G, s, t, k};
  return ret;
}

ds::ProblemInstance load_pace(char const *path) {
  std::ifstream f(path);
  if (f.fail()) throw std::invalid_argument(util::format("Failed to open file: %s", path));
  return read_pace(f);
}
}  // namespace readwrite

File: src/main/python/readwrite/pace.py
import networkx as nx
from typing import TextIO, Optional

__all__ = [
    'read_pace',
    'load_pace',
    'write_pace',
    'save_pace',
]


def read_pace(input: TextIO) -> tuple[nx.Graph, Optional[int], Optional[int], Optional[int]]:
    G = nx.Graph()
    given_n: Optional[int] = None
    given_m: Optional[int] = None
    given_c: Optional[int] = None
    given_s: Optional[int] = None
    given_t: Optional[int] = None
    given_k: Optional[int] = None
    all_colors: set[int] = set()

    for line in input.readlines():
        line = line.strip()
        if not line:
            continue
        if line.startswith('c'):
            continue  # ignore comments

        tokens = line.split()

        if tokens[0] == 'p':
            assert tokens[1] == 'ip', f'unexpected problem description: {tokens[1]}'

            given_n = int(tokens[2])
            given_m = int(tokens[3])
            given_c = int(tokens[4])
            given_s = int(tokens[5]) if len(tokens) >= 6 else None
            given_t = int(tokens[6]) if len(tokens) >= 7 else None
            given_k = int(tokens[7]) if len(tokens) >= 8 else None

        elif tokens[0] == 'v':
            v = int(tokens[1])
            colors = list(map(int, tokens[2:]))
            assert len(set(colors)) == len(colors), 'colors must be distinct'
            all_colors |= set(colors)
            G.add_node(v, color=colors)

        elif tokens[0] == 'e':
            assert len(tokens) == 4

            u = int(tokens[1])
            v = int(tokens[2])
            w = int(tokens[3]) if tokens[3].isdigit() else float(tokens[3])
            G.add_edge(u, v, weight=w)
        else:
            assert False, f'unexpected line: {line}'  # pragma: no cover

    assert given_n == G.number_of_nodes(), 'inconsistent n'
    assert given_m == G.number_of_edges(), 'inconsistent m'
    assert given_c == len(all_colors), 'inconsistent c'
    return G, given_s, given_t, given_k


def load_pace(path: str) -> tuple[nx.Graph, Optional[int], Optional[int], Optional[int]]:
    with open(path) as f:
        return read_pace(f)


def write_pace(output: TextIO, G: nx.Graph, s: Optional[int] = None, t: Optional[int] = None, k: Optional[int] = None, comments: list[str] = []) -> None:
    n = G.number_of_nodes()
    m = G.number_of_edges()
    assert all(isinstance(v, int) for v in G.nodes()), 'labels must be integers'

    c = len({v for vs in nx.get_node_attributes(G, 'color').values() for v in vs})

    for comment in comments:
        output.write(f'c {comment}\n')

    # p-line
    p_params = [n, m, c]
    if s is not None:
        p_params += [s]
        if t is not None:
            p_params += [t]
            if k is not None:
                p_params += [k]

    output.write(f'p ip {" ".join(map(str, p_params))}\n')

    # vertices
    for v in G.nodes():
        v_params = [v] + G.nodes[v]['color']
        output.write(f'v {" ".join(map(str, v_params))}\n')

    # edges
    for u, v in G.edges():
        output.write(f'e {u} {v} {G.edges[u, v]["weight"]}\n')


def save_pace(path: str, G: nx.Graph, s: Optional[int] = None, t: Optional[int] = None, k: Optional[int] = None, comments: list[str] = []) -> None:
    with open(path, 'w') as f:  # pragma: no cover
        write_pace(f, G, s, t, k, comments)
